manager/controllers/app/solve.go

```diff
@@ -59,10 +62,75 @@ func solve(env *datapath.Environment, datasets []datapath.DataInfo, log *zerolog
 }
 
 // perform basic checks before searching for a solution for a dataset
-func validateBasicConditions(env *datapath.Environment, log *zerolog.Logger) error {
+func validateBasicConditions(env *datapath.Environment, datasets []datapath.DataInfo, log *zerolog.Logger) error {
 	if len(env.Modules) == 0 {
 		log.Error().Msg(NoDeployedModules)
 		return errors.New(NoDeployedModules)
 	}
+	for i := range datasets {
+		dataset := &datasets[i]
+		if dataset.Context.Flow == "" || dataset.Context.Flow == taxonomy.ReadFlow {
+			if err := validateApplicationProtocol(env, dataset); err != nil {
+				log.Error().Err(err)
+				return err
+			}
+			if err := validateAssetProtocol(env, dataset); err != nil {
+				log.Error().Err(err)
+				return err
+			}
+		}
+	}
 	return nil
 }
+
+// create interface string to print in error messages
+func createInterfaceString(interfacePtr *taxonomy.Interface) string {
+	interfaceStr := string(interfacePtr.Protocol)
+	if interfacePtr.DataFormat != "" {
+		interfaceStr = interfaceStr + ", " + string(interfacePtr.DataFormat)
+	}
+	return interfaceStr
+}
+
+// check if any deployed module provides the requested read api by the application
+// return nil if such module exists, and an error if not
+func validateApplicationProtocol(env *datapath.Environment, dataset *datapath.DataInfo) error {
+	applicationInterfacePtr := dataset.Context.Requirements.Interface
+	for _, module := range env.Modules {
+		for _, capability := range module.Spec.Capabilities {
+			// check if the module capability matches the application protocol requirement
+			var capabilityInterfacePtr *taxonomy.Interface
+			if capability.API == nil {
+				continue
+			}
+			capabilityInterfacePtr = &taxonomy.Interface{Protocol: capability.API.Connection.Name, DataFormat: capability.API.DataFormat}
+			if match(capabilityInterfacePtr, applicationInterfacePtr) {
+				return nil
+			}
+		}
+	}
+	message := fmt.Sprintf("The requested interface (%s) is not supported by the deployed modules for dataset '%s'",
+		createInterfaceString(applicationInterfacePtr), dataset.Context.DataSetID)
+	return errors.New(message)
+}
+
+// check if any deployed module provides the connection to read the asset
+// return nil if such module exists, and an error if not
+func validateAssetProtocol(env *datapath.Environment, dataset *datapath.DataInfo) error {
+	assetConnection := dataset.DataDetails.Details.Connection.Name
+	assetDataformat := dataset.DataDetails.Details.DataFormat
+	assetInterfacePtr := &taxonomy.Interface{Protocol: assetConnection, DataFormat: assetDataformat}
+	for _, module := range env.Modules {
+		for _, capability := range module.Spec.Capabilities {
+			// check if the module capability matches the asset connection requirement
+			for _, readInterface := range capability.SupportedInterfaces {
```

Comment:
it is not necessarily read interface; the name is misleading.

Reply:
fixed